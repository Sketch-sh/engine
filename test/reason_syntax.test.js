const { evaluator: e } = require("../build/toplevel.js");
const objPath = require("object-path");

let count = (result, path, value) =>
  result.reduce((acc, phr_result) => {
    let valueAtPath = objPath.get(phr_result, path);
    if (typeof value === "function") {
      if (value(valueAtPath)) {
        return acc + 1;
      }
    } else if (valueAtPath === value) {
      return acc + 1;
    }
    return acc;
  }, 0);

let code = [
  // 0
  `print_endline("awesome")

  type say = Hello | Goodbye;

  fun
  | Hello => ()
  | Goodbye => ()
  | _ => ();

  let a = 1`,
  // 1
  `print_endline("Another stdout")`,
  // 2
  `let a = ref(1); print_int(a^);`,
  // 3
  `let rec factorial = (n) => n <= 0 ? 1 : n * factorial(n - 1); factorial(6);`,
  // 4
  `type tree =
  | Leaf
  | Node(int, tree, tree)
  
  let rec sum =
  fun
  | Leaf => 0
  | Node(value, left, right) => value + sum(left) + sum(right)
  
  let myTree =
  Node(
    1,
    Node(2, Node(4, Leaf, Leaf), Node(6, Leaf, Leaf)),
    Node(3, Node(5, Leaf, Leaf), Node(7, Leaf, Leaf)),
  )
  
  sum(myTree)`,
];

describe.each`
  nth  | phr_count | ok   | error | warning | stdout
  ${0} | ${4}      | ${4} | ${0}  | ${1}    | ${1}
  ${1} | ${1}      | ${1} | ${0}  | ${0}    | ${1}
  ${2} | ${2}      | ${2} | ${0}  | ${0}    | ${1}
  ${3} | ${2}      | ${2} | ${0}  | ${0}    | ${0}
  ${4} | ${4}      | ${4} | ${0}  | ${0}    | ${0}
`("$nth", ({ nth, phr_count, ok, error, warning, stdout }) => {
  let result = e.execute(code[nth]);

  test(`have ${phr_count} phrases`, () => {
    expect(result.length).toBe(phr_count);
  });

  test(`have ${ok} ok phrases`, () => {
    expect(count(result, "kind", "Ok")).toBe(ok);
  });

  test(`have ${error} error phrases`, () => {
    expect(count(result, "kind", "Error")).toBe(error);
  });

  test(`have ${warning} warnings`, () => {
    expect(
      count(result, "value.stderr", stderr => stderr.indexOf("Warning") > 0)
    ).toBe(warning);
  });

  test(`have ${stdout} stdout`, () => {
    expect(count(result, "value.stdout", stderr => stderr !== "")).toBe(stdout);
  });

  test("snapshot", () => {
    expect(result).toMatchSnapshot();
  });
});
