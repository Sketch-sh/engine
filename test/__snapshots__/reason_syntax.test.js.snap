// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`0 snapshot 1`] = `
Array [
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 0,
          "line": 1,
        },
        Object {
          "col": 24,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "awesome
",
      "value": "- : unit = ()
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 3,
        },
        Object {
          "col": 28,
          "line": 3,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "type say = Hello | Goodbye;
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 5,
        },
        Object {
          "col": 11,
          "line": 8,
        },
      ],
      "stderr": "File \\"\\", line 8, characters 2-5:
Warning 11: this match case is unused.
",
      "stdout": "",
      "value": "- : say => unit = <fun>
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 10,
        },
        Object {
          "col": 11,
          "line": 10,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "let a: int = 1;
",
    },
  },
]
`;

exports[`1 snapshot 1`] = `
Array [
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 0,
          "line": 1,
        },
        Object {
          "col": 31,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "Another stdout
",
      "value": "- : unit = ()
",
    },
  },
]
`;

exports[`2 snapshot 1`] = `
Array [
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 0,
          "line": 1,
        },
        Object {
          "col": 14,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "let a: ref(int) = {contents: 1};
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 16,
          "line": 1,
        },
        Object {
          "col": 29,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "1",
      "value": "- : unit = ()
",
    },
  },
]
`;

exports[`3 snapshot 1`] = `
Array [
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 0,
          "line": 1,
        },
        Object {
          "col": 60,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "let factorial: int => int = <fun>;
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 62,
          "line": 1,
        },
        Object {
          "col": 74,
          "line": 1,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "- : int = 720
",
    },
  },
]
`;

exports[`4 snapshot 1`] = `
Array [
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 0,
          "line": 1,
        },
        Object {
          "col": 25,
          "line": 3,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "type tree = Leaf | Node(int, tree, tree);
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 5,
        },
        Object {
          "col": 62,
          "line": 8,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "let sum: tree => int = <fun>;
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 10,
        },
        Object {
          "col": 3,
          "line": 15,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "let myTree: tree =
  Node(1, Node(2, Node(4, Leaf, Leaf), Node(6, Leaf, Leaf)),
   Node(3, Node(5, Leaf, Leaf), Node(7, Leaf, Leaf)));
",
    },
  },
  Object {
    "kind": "Ok",
    "value": Object {
      "loc": Array [
        Object {
          "col": 2,
          "line": 17,
        },
        Object {
          "col": 13,
          "line": 17,
        },
      ],
      "stderr": "",
      "stdout": "",
      "value": "- : int = 28
",
    },
  },
]
`;
